### ISSUE TITLE
Proposal: monad with `coerce`?

### ISSUE THREAD

**Issue by InfiniteEchoes** [formal-land/coq-of-rust#601]
``` Haskell
coerce :: M' 0 a → M a
coerce m = \x . let a = m x in (a, x)
```
As in [this paper](https://homepages.inf.ed.ac.uk/wadler/papers/marktoberdorf/baastad.pdf), I think many operations could be interesting to use, and moreover this might be very useful and more formal to resolve the mutual dependency issue

**Comment by clarus** [comment]
I think operators would not help to solve the dependencies issue, as operators are mostly syntactical sugar but does not change what we can express, or maybe I do not understand what you mean by mutual dependency issue.

**Comment by olympichek** [comment]
I second this. I don't understand what exactly do you mean by mutual dependency issue or how `coerce` can help with it. As for the paper and monadic operations, I would take a look at it later

**Comment by InfiniteEchoes** [comment]
> I second this. I don't understand what exactly do you mean by mutual dependency issue or how `coerce` can help with it. As for the paper and monadic operations, I would take a look at it later

Yes, for now I also don't take it seriously, since we haven't met anything that requires a serious investigation into the mutual dependency issue yet.
